### ISSUE TITLE
OpenSSL QUIC, seeing RESETs on short POSTs

### ISSUE THREAD

**Issue by icing** [openssl/openssl#23314]
### Description

In curl, we have a test cases for doing parallel POST requests to an nghttpx server. With request body sizes of 100KB, the OpenSSL client seems to work reliably. With requests bodies of just 10 bytes, there is a good chance that a stream is `RESET` and no response data, neither headers nor body, is being delivered to curl.

In that situation, `SSL_read_ex()` returns 0 and `SSL_get_stream_read_state()` returns `SSL_STREAM_STATE_RESET_REMOTE`. The error code retrieved via `SSL_get_stream_read_error_code()` is `0x100` (HTTP3_NO_ERROR). No QUIC data on that stream has otherwise been retrieved with `SSL_read_ex()`.

The nghttpx log lists the request headers and body data received and the response headers and data are logged as sent and acknowledged. It logs that it closed the stream with app error 0x100.

This looks like a race or a mistaken transition in OpenSSL QUIC stream state management.

If I can help with any printf debugging of the situation, please let me know.

**Comment by hlandau** [comment]
One possibility might be to try building with my QLOG PR and load the resulting trace into QVIS. https://github.com/openssl/openssl/pull/22037

If your code is available somewhere I may be able to find the time and try to reproduce this.

**Comment by icing** [comment]
@hlandau thanks for the offer. I am digging into the OpenSSL QUIC implementation and as I read the code, the handling of read on RESET streams seems to be wrong. My read:

`depack_do_frame_reset_stream()` handles the incoming reset packet, which switches the stream (in my tests) from `QUIC_RSTREAM_STATE_DATA_RECVD` to `QUIC_RSTREAM_STATE_RESET_RECVD`.

When doing a read on such a stream, this ends in `quic_read_actual()` which does a check if the stream is valid for reading by calling `quic_validate_for_read()`. That function returns FALSE on stream in state `QUIC_RSTREAM_STATE_RESET_RECVD` and sets the error to `SSL_R_STREAM_RESET`.

Which means, no matter how much data is pending in OpenSSL to be read, once a RESET frame has been ingested, the reading will always fail.

This is not how RESETs of a stream is supposed to work. An application must be able to read the data that arrived *before* the RESET, no matter at which time this happens, for reliable behaviour.

**Comment by hlandau** [comment]
> @hlandau thanks for the offer. I am digging into the OpenSSL QUIC implementation and as I read the code, the handling of read on RESET streams seems to be wrong. My read:
> 
> `depack_do_frame_reset_stream()` handles the incoming reset packet, which switches the stream (in my tests) from `QUIC_RSTREAM_STATE_DATA_RECVD` to `QUIC_RSTREAM_STATE_RESET_RECVD`.
> 
> When doing a read on such a stream, this ends in `quic_read_actual()` which does a check if the stream is valid for reading by calling `quic_validate_for_read()`. That function returns FALSE on stream in state `QUIC_RSTREAM_STATE_RESET_RECVD` and sets the error to `SSL_R_STREAM_RESET`.
> 
> Which means, no matter how much data is pending in OpenSSL to be read, once a RESET frame has been ingested, the reading will always fail.
> 
> This is not how RESETs of a stream is supposed to work. An application must be able to read the data that arrived _before_ the RESET, no matter at which time this happens, for reliable behaviour.

This was the deliberate design of the reset handling. Protocols based on QUIC should use a reset when they no longer require transmission of a stream's data. As such, there is no particular delivery or reception guarantee once a reset occurs.

If https://datatracker.ietf.org/doc/draft-ietf-quic-reliable-stream-reset/ becomes an RFC and we implement in the future this may change. But this is explicitly a change from standard QUIC.

This is all based on my interpretation of QUIC as intended to be used. If you think there's something wrong with the implementation rationale here, I'm happy to discuss it...

Can you cite a standard/I-D protocol built on QUIC which requires handling resets in this way?

**Comment by t8m** [comment]
The STREAM_RESET can be interpreted as "the data I've sent so far are no longer relevant".

**Comment by icing** [comment]
"Can be", but that is not how HTTP/3 servers do it. HTTP/3 servers use RESET with app error `H3_NO_ERROR (0x0100)` to indicate to the client that they are done with the response on the stream.

Example: a client make a large upload to a resource that need authentication, the server decides to answer with a 401 and RESETs the stream after sending the response, since it does not want any further upload data.

The current OpenSSL design would make it a roll of the dice if the client actually sees the 401 response or just the RESET, depending on when it comes and what it has read already. This is clearly not acceptable. Note that HTTP/2 RESETs work the same way.

Your interpretation may be possible, but it is not interoperable.

**Comment by hlandau** [comment]
RFC 9114:

>A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When the server does not need to receive the remainder of the request, it MAY abort reading the [request stream](https://www.rfc-editor.org/rfc/rfc9114.html#request-streams), send a complete response, and cleanly close the sending part of the stream. The error code [H3_NO_ERROR](https://www.rfc-editor.org/rfc/rfc9114.html#H3_NO_ERROR) SHOULD be used when requesting that the client stop sending on the [request stream](https://www.rfc-editor.org/rfc/rfc9114.html#request-streams).

This does not say that a server can reset its response. It says that the server can send STOP_SENDING to the client which will force the client to respond with STREAM_RESET on the client send part (i.e., the *request* component of the stream). Note that resets on the send and receive parts of a bidirectional stream are independent.

Per the above, a server may respond with a response to a request before reading the entire request. This does **not** say that the response part of a stream can be reset and my reading is that this is not in compliance with RFC 9114.

**Comment by icing** [comment]
You are a bit late to the party here in telling server QUIC implementations how to behave, don't you think?

**Comment by hlandau** [comment]
> In respect to your interpretation above, why do you think `H3_NO_ERROR` was named that way if it invalidates the response that was sent according to your interpretation? Clearly, the client is expected to see the data that was transmitted before the RESET. And this is the only thing that makes sense here...

`H3_NO_ERROR` is entirely valid for use &mdash; in `STOP_SENDING`. There is no contradiction here.

**Comment by hlandau** [comment]
To be clear &mdash; bad implementations are a fact of life and it's not as though we don't have compatibility options for interop with bad implementations in our TLS implementation. If people want the ability to access already received data after a reset I'm not intrinsically against this as a feature and there are probably some valid use cases (aside from interop with faulty HTTP/3 servers). An API would need to be designed. This would in any case be a feature not a bug fix.

Nonetheless, the handling of resets in this regard is expressly in alignment with RFC 9000 and the design intention of the QUIC stream reset feature in my view. I nonetheless remain open to discussion on this point.

It is not particularly surprising to me that the history of evolution of QUIC and its implementations has led to HTTP/3 implementations which are not entirely compliant with the contemporary RFC, but that does not make them compliant. It seems like it would be productive to raise bugs with affected HTTP/3 servers.

**Comment by icing** [comment]
To summarise where we are at:

1. From QUIC protocol perspective, 19.4 says: "A receiver of RESET_STREAM can discard any data that it already received on that stream." OpenSSL's implementation will really discard *all* data that has not been read by the application yet. Even data that has been ACKed to the server.

2. Other QUIC client implementations (ngtcp2/quiche) interpret this by *not* discarding ACKed data.

3. Since there is need to sending RESETs where data in flight is intended to be read by the QUIC application, Marten Seemann initiated https://datatracker.ietf.org/doc/draft-ietf-quic-reliable-stream-reset/ in order to resolve possible differences in interpretation. But especially for the handling of unacked packets pending delivery.

As application, `curl` sees the differences in behaviour and interoperability for the QUIC implementation used. While ngtcp2 and quiche give us reliable interop, OpenSSL's Quic fails randomly because of, you say "bad servers".

I have not determined how many "bad servers" are out there. I see that using nghttpx (ngtcp2+nghttp3) in proxy mode.

**Comment by hlandau** [comment]
I think those bullet points are a fair summary, though "bad servers" in this case means in contradiction to RFC 9114.

Do you have the precise versions of the ngtcp2 and nghttp3 packages you are using? Thanks.

**Comment by icing** [comment]
Talking with @tatsuhiro-t. He will see what/if he wants to change in this scenario. I'll report when I get an update.

**Comment by icing** [comment]
Update: this looks like a bug in nghttpx and @tatsuhiro-t is working on it. I propose to leave this issue open until I verified the fix in my setup?

**Comment by hlandau** [comment]
Absolutely. Thanks for taking the time to investigate this! And let me know if you hit any other issues.

**Comment by mattcaswell** [comment]
Looks like the fix was merged in nghttp2 (https://github.com/nghttp2/nghttp2/pull/2036).

So can this be closed?

**Comment by MikeBishop** [comment]
While I'm late to this discussion, I just wanted to add my USD 0.02 for clarity.  @hlandau was exactly correct about early responses:

- The H3 server aborts read on the stream (which at the QUIC layer produces a STOP_SENDING) to indicate that it does not require any further bytes from the client to generate its response, even though more bytes remain unread or in flight.
- When the application layer aborts read on a stream, it is entirely correct for the QUIC layer to discard all bytes which have not yet been read, regardless of whether the packets containing them have been acknowledged. The application has explicitly said it does not require and will not consume further data.
- When the client receives the QUIC STOP_SENDING frame, what it does is bookkeeping -- if any bytes remain unacknowledged, it sends a RESET_STREAM instead of continuing to retransmit, because it knows the data is not needed. The sender owns its sending stream, and must be the one to actually complete the requested state transition.
- The state of the client-to-server stream is independent from the server-to-client stream. A STOP_SENDING from the server does not affect the state of the stream from the server to the client, which can continue transmitting data and being read by the client normally.  The response should be consumed (if the stream terminates cleanly, i.e. with a FIN) or discarded (if the stream terminates prematurely with a RST_STREAM) according to the state progression of the stream from server to client.

@icing was correctly describing a different situation, the receipt of an unsolicited RESET_STREAM when reading has not been aborted by the application layer. At that point, handling of received-but-unread data is explicitly implementation-dependent:

> Sending a RESET_STREAM means that an endpoint cannot guarantee delivery of stream data; however, there is no requirement that stream data not be delivered if a RESET_STREAM is received. An implementation MAY interrupt delivery of stream data, discard any data that was not consumed, and signal the receipt of the RESET_STREAM. A RESET_STREAM signal might be suppressed or withheld if stream data is completely received and is buffered to be read by the application. If the RESET_STREAM is suppressed, the receiving part of the stream remains in "Data Recvd".
>
>Once the application receives the signal indicating that the stream was reset, the receiving part of the stream transitions to the "Reset Read" state, which is a terminal state.

It's true that many QUIC implementations choose to err on the side of allowing the application to read any data that was previously received and only indicating the RESET_STREAM when the application reaches a byte that was unavailable. However, that is not behavior on which the client can depend; hence the work on the RESET_STREAM_AT frame which guarantees delivery of data up to a particular offset and indicating the RESET at that point or later.

It appears that nghttp2 was incorrectly resetting the server-to-client stream after sending a response, causing the second case to occur in parallel with the first.

**Comment by hlandau** [comment]
The above is a good summary of some of the relevant mechanisms here.

**Comment by icing** [comment]
Thanks @MikeBishop for your summary. In this case, what turned out to have happened is:

 * backend H2 closes stream with DATA frame FIN flag set.
 * nghttpx converted that in STOP_SENDING and STREAM with fin bit set
 * nghttpx then, in come mistaken attempt to cleanup "dangling" streams, sent a RESET_STREAM in addition after the  previous ones with - intentionally or not - the previous packets ACKed.

This is where client implementation differed then, as I suspect. ngtcp2 and quiche ignore the RESET_STREAM on a stream they already handled in full, while OpenSSL clears all stream data that are pending for delivery to the application.

Now, nghttpx has reverted the change that led to the RESET, so this is historical. But I thought I add the complete picture here.
